misc/python/materialize/cli/ci_flake_detect.py

```diff
@@ -0,0 +1,116 @@
+
+import argparse
+import re
+import sys
+from typing import Any, List
+
+import junit_xml
+import requests
+
+from materialize import ci_util
+
+CI_RE = re.compile("^ci-regexp: (.*)$", flags=re.MULTILINE)
+FAILURE_RE = re.compile("(panicked|Out of memory)")
+
+
+class Flake:
+    def __init__(self, regex: str, info: Any):
+        self.regex = re.compile(regex)
+        self.info = info
+
+
+def main(argv: List[str]) -> int:
+    parser = argparse.ArgumentParser(
+        prog="ci-flake-detect",
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        description="""
+ci-flake-detect detects known flakes using the log files and open Github issues
+on Materialize repository.""",
+    )
+
+    parser.add_argument(
+        "--name", required=True, help="identifier to use for creating junit xml file"
+    )
+
+    parser.add_argument("log_files", nargs="+", help="log files to search in")
+    args = parser.parse_args()
+
+    annotate_known_failures(args.log_files, args.name)
+    return 0
+
+
+def annotate_known_failures(log_files: List[str], name: str) -> None:
+    """
+    Called when a test failure occurred, check if it is a known flaky CI failure by
+    checking the existing Github issues.
+    """
+
+    failure_lines = get_failure_lines(log_files)
+
+    if not failure_lines:
+        print("No failures found in logs")
+        return
+
+    flakes = get_flakes_from_github()
+
+    junit_suite = junit_xml.TestSuite("Flakes")
+
+    junit_report = ci_util.junit_report_filename(f"{name}_flakes")
+    for i, line in enumerate(failure_lines):
+        test_case = junit_xml.TestCase(f"Flake {i}", "Flakes")
+        for flake in flakes:
+            match = flake.regex.search(line)
+            if match:
+                test_case.add_error_info(
+                    message=f"Known failure in logs: {flake.info['title']} (#{flake.info['number']}): {flake.info['html_url']}",
+                    output=line,
+                )
+                break
+        else:
+            test_case.add_error_info(message="Unknown failure in logs", output=line)
+        junit_suite.test_cases.append(test_case)
+
+    junit_report = ci_util.junit_report_filename(f"{name}_flakes")
+    with junit_report.open("w") as f:
+        junit_xml.to_xml_report_file(f, [junit_suite])
+    ci_util.upload_junit_report(f"{name}_flakes", junit_report)
+
+
+def get_failure_lines(log_files: List[str]) -> List[str]:
+    failure_lines = []
+    for log_file in log_files:
+        with open(log_file) as f:
+            for line in f:
+                match = FAILURE_RE.search(line)
+                if match:
+                    failure_lines.append(line)
+    return sorted(set(failure_lines))
+
+
+def get_flakes_from_github() -> list[Any]:
+    flakes_json = requests.get(
+        f"https://api.github.com/search/issues?q=type:repo:MaterializeInc/materialize%20type:issue%20state:open%20label:ci-flake"
+    ).json()
```

Comment:
I guess you need to check for an 200 HTTP error code and , for the time being, fail the CI step if one is not received. We can later just log an error once we have seen how this works in practice.